### ISSUE TITLE
Problem parsing a compound expression in [[ ]]

### ISSUE THREAD

**Issue by bryder** [BashSupport-Pro/bashsupport-pro#65]
Hi,

```bash
if [[ ${BASH_VERSINFO[0]} -lt 4 || ( ${BASH_VERSINFO[0]} -eq 4 && ${BASH_VERSINFO[1]} -lt 4 )  ]]; then
  echo "${BASH_SOURCE[0]}: **ERROR** Only bash version 4.4 and later is known to work"
  return 1
fi
```
causes these errors

* ')' expected, got '-eq'
* ';', PIPE_AMP &, PIPE_AND_AND &&, PIPE_OR_OR ||, PIPE_PIPE, '\n' or '|&' expected, got ')'

This is with IntelliJ 2020.3 and 1.5.14.203 and 1.5.14.2.203 (the preview).

Thanks for the great plugin!
Bill

**Comment by bryder** [comment]
As an aside it seems to be related to the positioning of the parentheses.

This gives the same errors - but they move to the first parenthesised expression.

```bash
if [[ ( $maj -lt 4 ) || ( $maj -eq 4 && $min -lt 4 )  ]]; then
  echo "${BASH_SOURCE[0]}: **ERROR** Only bash version 4.4 and later is known to work"
  return 1
fi
```

So you get
* ')' expected, got '-lt'
* ';', PIPE_AMP &, PIPE_AND_AND &&, PIPE_OR_OR ||, PIPE_PIPE, '\n' or '|&' expected, got ')'

No errors are indicated for the expression following the '||' in this case.

And I do realise I don't need to put the && expressions in parentheses. I just like to make it visually explicit :-)

**Comment by jansorg** [comment]
Thanks, as you noticed this is most likely caused by grouped test expressions. I‘ll take a closer look and investigate soon.

2.0beta already has improved parsing of test expressions, so the fix might end up in the next major version instead of 1.5.x. Would that be okay for you?

**Comment by bryder** [comment]
I can wait for 2.0 thanks.

**Comment by jansorg** [comment]
I fixed this for 2.0.

The latest preview already has the fix, just in case that you'd like to use it:
> Changelog and install instructions at:
https://plugins.jetbrains.com/plugin/13841-bashsupport-pro/next-version

**Comment by bryder** [comment]
Great thanks! I tried it out  on 2020.3

and got these errors. I wonder if I should try 2021.1 again? 

Since the errors are a bit strange - CLOSE_QUOTE expected got '"'

Same for the RIGHT_CURLY

```text
<expression> expected, got '||'
```` 
for this
```bash
if [[ ! -v "${variable_name}" || ${#blah[@]} -gt 10  ]] ; then
   echo 'in then'
fi
```

And
```text
$(, $((, $[ or CLOSE_QUOTE expected, got '"'
and
$(, $((, $[, CLOSE_QUOTE or '[' expected
```
for this
```bash
echo "stuff: $(declare -p blah )"
```
that one is quite bad because it hasn't terminated the string to it has flow on effects

And I got 
```text
$(, $((, $[ or RIGHT_CURLY expected, got '}'
````
For
```bash
printf "%s" "${usage_string[@]#    }"
```

**Comment by bryder** [comment]
I get the same errors on 2021.1 

```text
IntelliJ IDEA 2021.1 EAP (Ultimate Edition)
Build #IU-211.5538.20, built on February 5, 2021
IntelliJ IDEA EAP User
Expiration date: March 7, 2021
Runtime version: 11.0.10+8-b1304.1 amd64
VM: Dynamic Code Evolution 64-Bit Server VM by JetBrains s.r.o.
Linux 4.14.170-weta-20200213
GC: ParNew, ConcurrentMarkSweep
Memory: 8012M
Cores: 12
Registry: directory.access.checker.enabled=false
Non-Bundled Plugins: YAML/Ansible support (0.11.2), com.andrewbrookins.wrap_to_column (1.4.0), com.jetbrains.plugins.ini4idea (211.5538.10), de.dieploegers.develop.idea.shellfilter (3.0.3), emacsIDEAs (1.4.9), eu.osimowicz.plugins.intellij.CodeBlocksSorter (0.1.2), net.seesharpsoft.intellij.plugins.csv (2.14.3), org.jetbrains.plugins.localization (211.5538.10), org.jetbrains.plugins.rest (211.5538.10), com.intellij.lang.puppet (211.5538.10), org.toml.lang (0.2.140.3644-203), Builder Generator (1.1.8), com.intellij.kubernetes (211.5538.10), String Manipulation (8.10.191.000.0), com.intellij.plugin.adernov.powershell (2.0.5), com.perl5 (211-SNAPSHOT), pro.bashsupport (2.0.0.211-beta4), Dart (211.5538.10), org.jetbrains.plugins.ruby (211.5538.10), Pythonid (211.5538.22), org.jetbrains.plugins.go (211.5538.10)
Kotlin: 211-1.4.21-release-IJ5538.2
Current Desktop: KDE
```

**Comment by jansorg** [comment]
@bryder Thanks! Unfortunately these are bugs, but unrelated to the fix of the original issue. 
2.0 has a new lexer and improved support for conditional expressions and the new issues are regressions compared to 1.5. I've fixed all of them for the next beta version.

**Comment by bryder** [comment]
Hi.
These all seem to be fixed in 2.0.0.211-beta6-eap
Excellent!
